File: python/connection-layer/schedule.py
"""Block proposer schedule — deterministic round-robin over the validator set.

The proposer for height H is validators_sorted_by_pubkey[H % N]. Every hub computes
the same schedule, so equal-height forks resolve deterministically: the block sealed
by the scheduled proposer wins; if neither (or both) match, the lower hash wins.
"""

from __future__ import annotations

from typing import Any


def sorted_validator_pubkeys() -> list[str]:
    from validators import known_validators

    return sorted(known_validators().keys())


def proposer_pubkey_for_height(height: int) -> str:
    pubkeys = sorted_validator_pubkeys()
    if not pubkeys:
        return ""
    return pubkeys[height % len(pubkeys)]


def is_local_turn(height: int) -> bool:
    from validators import validator_info

    scheduled = proposer_pubkey_for_height(height)
    return not scheduled or scheduled == validator_info().public_key


def schedule_snapshot(next_height: int, count: int = 5) -> list[dict[str, Any]]:
    from validators import known_validators

    registry = known_validators()
    rows = []
    for height in range(next_height, next_height + count):
        pubkey = proposer_pubkey_for_height(height)
        info = registry.get(pubkey)
        rows.append(
            {
                "height": height,
                "proposer_pubkey": pubkey,
                "validator_id": info.validator_id if info else None,
                "hub_url": info.hub_url if info else None,
            }
        )
    return rows


def tiebreak_wins(candidate_tip: dict[str, Any], local_tip: dict[str, Any]) -> bool:
    """True if candidate tip beats local tip at the same height."""
    height = int(candidate_tip.get("index", 0))
    scheduled = proposer_pubkey_for_height(height)
    cand_signer = str((candidate_tip.get("proof") or {}).get("validator_pubkey", ""))
    local_signer = str((local_tip.get("proof") or {}).get("validator_pubkey", ""))
    if scheduled:
        if cand_signer == scheduled and local_signer != scheduled:
            return True
        if local_signer == scheduled and cand_signer != scheduled:
            return False
    return str(candidate_tip.get("hash", "")) < str(local_tip.get("hash", ""))

File: rust/crates/noetis-chain/src/schedule.rs
//! Deterministic block proposer schedule — port of Python `schedule.py`.

use serde_json::Value;

/// Proposer for height H is `sorted_pubkeys[H % N]`.
pub fn proposer_pubkey_for_height(height: i64, sorted_pubkeys: &[String]) -> String {
    if sorted_pubkeys.is_empty() {
        return String::new();
    }
    let index = (height.rem_euclid(sorted_pubkeys.len() as i64)) as usize;
    sorted_pubkeys[index].clone()
}

pub fn is_local_turn(height: i64, sorted_pubkeys: &[String], local_pubkey: &str) -> bool {
    let scheduled = proposer_pubkey_for_height(height, sorted_pubkeys);
    scheduled.is_empty() || scheduled == local_pubkey
}

/// Python `tiebreak_wins` — candidate beats local tip at equal height when it
/// was sealed by the scheduled proposer (falling back to lower hash).
pub fn tiebreak_wins(candidate_tip: &Value, local_tip: &Value, sorted_pubkeys: &[String]) -> bool {
    let height = candidate_tip.get("index").and_then(Value::as_i64).unwrap_or(0);
    let scheduled = proposer_pubkey_for_height(height, sorted_pubkeys);
    let cand_signer = candidate_tip
        .get("proof")
        .and_then(|p| p.get("validator_pubkey"))
        .and_then(Value::as_str)
        .unwrap_or("");
    let local_signer = local_tip
        .get("proof")
        .and_then(|p| p.get("validator_pubkey"))
        .and_then(Value::as_str)
        .unwrap_or("");
    if !scheduled.is_empty() {
        if cand_signer == scheduled && local_signer != scheduled {
            return true;
        }
        if local_signer == scheduled && cand_signer != scheduled {
            return false;
        }
    }
    let cand_hash = candidate_tip.get("hash").and_then(Value::as_str).unwrap_or("");
    let local_hash = local_tip.get("hash").and_then(Value::as_str).unwrap_or("");
    cand_hash < local_hash
}
